File: 0086 Partition List/solution.py
#! python3

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        nL1 = L1 = ListNode(-1)
        nL2 = L2 = ListNode(-2)
        
        while head != None:
            if head.val < x:
                L1.next = head
                L1 = L1.next
            else:
                L2.next = head
                L2 = L2.next
            head = head.next
        
        L2.next = None
        L1.next = nL2.next
        
        return nL1.next

def createLinkedList(nums):
    if not nums or len(nums) == 0:
        return None
    head = p= ListNode(nums[0])
    for n in nums[1:]:
        p.next = ListNode(n)
        p = p.next
    return head

def printAndClear(head):
    while head:
        print(head.val, end='')
        if head.next:
            print("->", end='')
        d, head = head, head.next
        del d




if __name__ == "__main__":
    head = createLinkedList([1,4,3,2,5,2])
    res = Solution().partition(head, 3)
    printAndClear(res)

File: 0086 Partition List/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *l1 = new ListNode(0);
        ListNode *l2 = new ListNode(0);
        ListNode *p = l1, *q = l2;

        for (ListNode *t = head; t; t = t->next)
            if (t->val < x)
            {
                p->next = t;
                p = t;
            }
            else
            {
                q->next = t;
                q = t;
            }

        p->next = l2->next;
        q->next = 0;

        return l1->next;
    }
};

ListNode *createLinkedList(vector<int>::iterator beg, vector<int>::iterator end)
{
    ListNode *head = beg == end ? NULL : new ListNode(*beg ++);
    for (ListNode *cur = head; beg != end; cur = cur->next)
        cur->next = new ListNode(*beg++);
    return head;
}

void printAndClear(ListNode *head)
{
    while (head)
    {
        cout << head->val;
        if (head -> next) cout << "->";
        ListNode *del = head;
        head = head->next;
        delete del;
    }
    cout << endl;
}

int main(){
    vector<int> vec{1, 4, 3, 2, 5, 2};
    ListNode *head = createLinkedList(vec.begin(), vec.end());
    Solution s;
    ListNode *res = s.partition(head, 3);
    printAndClear(res);
    return 0;
}